### ISSUE TITLE
All text is in Chinese after update

### ISSUE THREAD

**Issue by DomPrice** [KSP-CKAN/CKAN#3244]
I updated CKAN to the current version and it changed the program language to chinese. I cannot change it back to english and do not know where to fix this. 

![Capture](https://user-images.githubusercontent.com/857194/102736940-25f2ad00-4314-11eb-9ed5-5fc42c954e94.PNG)

**Comment by HebaruSan** [comment]
Open the second menu, first option to get to the settings, then there's a language dropdown towards the bottom. Change it and restart.

**Comment by DomPrice** [comment]
I was able to find the config file under AppData after some digging and change it there. I don't know how it got changed in the first place. :/

**Comment by HebaruSan** [comment]
Yeah somebody reported the same on Discord a couple of days ago. No clear pattern yet.

**Comment by DomPrice** [comment]
Wierd.... thanks for looking into it though!

**Comment by HebaruSan** [comment]
It's interesting that this is linked to CKAN upgrades. That vaguely suggests a timing or race condition in which the old CKAN has not finished saving its settings somehow when the new CKAN is started.

But after auditing the config saving/loading/parsing and language setting code, I do not see any way that it could randomly get set to `zh-CN` in a typical `en-US` or `en-GB` environment. If the `config.json` file doesn't exist, it falls back to the old registry if applicable, or instantiates a new object in which `config.Language` would be `null`, which would cause it to use the culture inherited from the environment.

Is there some way that `Thread.CurrentThread.CurrentUICulture` could be set to Chinese by mono at startup on a non-Chinese PC? Very doubtful.

Maybe `Autofac` is to blame somehow? It is involved in loading the settings, and the only piece of this that I am not confident I understand. The code in `ServiceLocator` is... weird, a little awkward, hard to follow. Not inconceivable that there would be bugs there, I suppose.

**Comment by HebaruSan** [comment]
This was also reported on Discord on Dec 11:

![screenshot](https://cdn.discordapp.com/attachments/601456752906469388/786992790587441192/CKAN_v1.29.2_-_KSP_1.8.1.2694_--_D__Games_Kerbal_Space_Program_11.12.2020_18_28_10.png)

**Comment by DasSkelett** [comment]
Another case reported on the r/KSP Discord. Only information I was able to get was

> Not sure.  I tried to install another mod when I received an error message. I restarted the application and noticed the language change

**Comment by HebaruSan** [comment]
Nobody ever seems to get German by accident! :frowning:

**Comment by HebaruSan** [comment]
Finally some variety:

![image](https://user-images.githubusercontent.com/1559108/110532628-b5715600-80e2-11eb-901d-0898072cc689.png)

**Comment by vinix38** [comment]
This comment might be useless, but anyway: it seems that the language gets switched to the last one on the list.
![image](https://user-images.githubusercontent.com/68441743/110922174-ac20fd00-831f-11eb-9a5d-46d160e039fa.png)
People got Chinese, until French was added at the end of the list and now people get French (I assume Toke got French, since CKAN is not translated in Spanish).
I have absolutely no idea what the reason is code-wise, but it seems to be related

**Comment by HebaruSan** [comment]
> This comment might be useless

No, it's a good insight, you might be completely right about that.
So if we can't figure out a full fix, a workaround would be to rearrange the list to put en-GB or en-US at the end. :grin:

**Comment by HebaruSan** [comment]
In my previous audit of the i18n code, I looked at how the setting is handled on startup. I've checked that code again a couple of times since, and it still looks rock-solid to me.

That leaves the settings window language dropdown. For this to be the cause, each user reporting this must have opened and closed the settings window in the session before the problem started, without having chosen a language before, and just didn't mention it. I don't think that's too much of a stretch, since they'd have no reason to remember having done that.

Loading the combobox:

https://github.com/KSP-CKAN/CKAN/blob/303c47c318ee99768b271f901a9f9a226bce034f/GUI/Dialogs/SettingsDialog.cs#L106-L114

Setting `SelectedIndex` calls the change handler, which has a null reference bug hidden in it:

https://github.com/KSP-CKAN/CKAN/blob/303c47c318ee99768b271f901a9f9a226bce034f/GUI/Dialogs/SettingsDialog.cs#L544-L547

That will throw if `LanguageSelectionComboBox.SelectedItem` is null, which it will be if the user has never selected a language:

> The object that is the currently selected item or [null if there is no currently selected item](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.combobox.selecteditem?view=netframework-4.7.2).

But before we even get to that point, there may be some funny business with `SelectedIndex`. [The `FindStringExact` call will return -1 if `config.Language` isn't in the language list](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.combobox.findstringexact?view=netframework-4.7.2). But `SelectedIndex` only supports -1 as a value if `FormattingEnabled ` is true:

> SelectedIndex, SelectedValue, and FormattingEnabled are related as follows:
> 
> - [If FormattingEnabled is false, SelectedIndex will not be set to -1 when SelectedValue is blank.](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.combobox.selectedindex?view=netframework-4.7.2)
> - If FormattingEnabled is true, SelectedIndex will be set to -1 when SelectedValue is blank.

`FormattingEnabled` is not set:

https://github.com/KSP-CKAN/CKAN/blob/303c47c318ee99768b271f901a9f9a226bce034f/GUI/Dialogs/SettingsDialog.Designer.cs#L556-L565

And the default is `false`!

> true if formatting of the DisplayMember property is enabled; otherwise, false. [The default is false.](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.listcontrol.formattingenabled?view=netframework-4.7.2)

That means our combobox **isn't** using an index of -1 to represent null, but the documentation doesn't say what it represents instead in that case.

In Python, `array[-1]` refers to the last element in the array. Meanwhile here we have a -1 value of unknown meaning, _and_ a value is being set to the last element of a list? Maybe that's not just a coincidence...

**Comment by HebaruSan** [comment]
Purging the language field from my `config.json`, I get `en-US` when I first open the settings:

![image](https://user-images.githubusercontent.com/1559108/190527200-ed570389-253d-413b-bc8d-6ab0ced55213.png)

So that's being pulled from my environment. Maybe the people having this problem had a locale setting that isn't in our list?

**Comment by HebaruSan** [comment]
Purging the JSON again and temporarily commenting `en-US` out of `Utilities.AvailableLanguages` gives me a blank combobox:

![image](https://user-images.githubusercontent.com/1559108/190527496-564f97f9-fb76-49e2-99d4-ee4c9226e18a.png)

After that, `config.json` says:

```json
  "Language": null,
```

... and that seems to be stable. With a null value in the JSON, the combobox stays empty, and the JSON value stays null.

**Comment by HebaruSan** [comment]
The JSON file is pretty well protected from corruption:

https://github.com/KSP-CKAN/CKAN/blob/303c47c318ee99768b271f901a9f9a226bce034f/Core/Configuration/JsonConfiguration.cs#L182-L203
